kubernetes/metadata/resource.go

```diff
@@ -167,23 +172,27 @@ func (r *Resource) GenerateK8s(kind string, obj kubernetes.Resource, options ...
 }
 
 func generateMapSubset(input map[string]string, keys []string, dedot bool) mapstr.M {
+	var matched bool
 	output := mapstr.M{}
 	if input == nil {
 		return output
 	}
 
 	for _, key := range keys {
-		value, ok := input[key]
-		if ok {
-			if dedot {
-				dedotKey := utils.DeDot(key)
-				_, _ = output.Put(dedotKey, value)
-			} else {
-				_ = safemapstr.Put(output, key, value)
+		for label, value := range input {
+			matched, _ = regexp.MatchString(key, label)
```

Comment:
My concern is that in general this change might affect the performance of the library since it will execute more cycles even if there is no regexp provided. Would you consider adding this behind an advanced setting just in case we hit any performance issues?

For the specific line, if I'm not mistaken it would be more performant if you call https://pkg.go.dev/regexp#Compile first outside of the loop and then call the [MatchString](https://pkg.go.dev/regexp#Regexp.MatchString) on the result. With this you only take the compile's cost once outside of the loop and then calling the less expensive match on the result. See the sample of https://pkg.go.dev/regexp#Regexp.MatchString for a more complete example.

Reply:
@ChrsMark regarding your comment above, the description and need for this comes from https://github.com/elastic/elastic-agent/issues/3346 enhancement request. 

I will try to add it behind a flag for performance concerns

For examples please see `kubernetes/metadata/resource_test.go` where I tested those 

- Validating wildcards
- Validating ?
- Validating start of a string
- Validating end of a string
- Exact matches
- Dedot Validation

Reply:
@ChrsMark for the two comments above:

For the flag to conditionally include/exclude the regex I added https://github.com/elastic/elastic-agent-autodiscover/pull/60/files#diff-c707b2fb5fad494cfb847530f0fd100719b6c3b4fd66706153248d6a46bf46fcR50-R51

For the https://pkg.go.dev/regexp#Compile now: This validates whether the regex is satisfied and returns the regex. Then we will need again to run the `regexp.MatchString` to take the result back. So this will add some more processing. 

I guess with the flags (UseRegexInclude , UseRegexExclude) we are ok as far as the concern for extra processing. What do you think?

Reply:
I update my comment as I used compile afterall @ChrsMark .
`Because the pattern only needs to be compiled once, this is more efficient.`!
From [here](https://blog.logrocket.com/deep-dive-regular-expressions-golang/)

Reply:
> For examples please see kubernetes/metadata/resource_test.go where I tested those

Tests are great but does not really help to get the high level purpose of the PR. I would just provide some examples in the PR's description with Pod's annotations/labels and various settings so as to illustrate the the outcome of this patch.

In any case, feel free to add those here in the description or those can be covered in the PR that will update the docs. It's up to you.